File: Objective-C/Neocom/Neocom/NCLabel.m
//
//  NCLabel.m
//  Neocom
//
//  Created by Artem Shimanski on 14.11.16.
//  Copyright © 2016 Artem Shimanski. All rights reserved.
//

#import "NCLabel.h"

@interface NCLabel()
@property (nonatomic, assign) CGFloat pointSize;
- (CGFloat) fontSizeWithContentSizeCategory:(UIContentSizeCategory) contentSizeCategory;
@end

@implementation NCLabel

- (void) awakeFromNib {
	[super awakeFromNib];
	self.pointSize = self.font.pointSize;
	[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(didChangeContentSizeCategory:) name:UIContentSizeCategoryDidChangeNotification object:nil];
	self.font = [self.font fontWithSize:[self fontSizeWithContentSizeCategory:[[UIApplication sharedApplication] preferredContentSizeCategory]]];
}

- (void) dealloc {
	[[NSNotificationCenter defaultCenter] removeObserver:self];
}

/*
// Only override drawRect: if you perform custom drawing.
// An empty implementation adversely affects performance during animation.
- (void)drawRect:(CGRect)rect {
    // Drawing code
}
*/

#pragma mark - Private

- (void) didChangeContentSizeCategory:(NSNotification*) not {
	self.font = [self.font fontWithSize:[self fontSizeWithContentSizeCategory:[[UIApplication sharedApplication] preferredContentSizeCategory]]];
	[self invalidateIntrinsicContentSize];
}

- (CGFloat) fontSizeWithContentSizeCategory:(UIContentSizeCategory) contentSizeCategory {
	CGFloat p = self.pointSize;
	if ([contentSizeCategory isEqualToString:UIContentSizeCategoryExtraSmall])
		p -= 2;
	else if ([contentSizeCategory isEqualToString:UIContentSizeCategorySmall])
		p -= 1;
	else if ([contentSizeCategory isEqualToString:UIContentSizeCategoryLarge])
		p += 1;
	else if ([contentSizeCategory isEqualToString:UIContentSizeCategoryExtraLarge])
		p += 3;
	else if ([contentSizeCategory isEqualToString:UIContentSizeCategoryExtraExtraLarge])
		p += 5;
	else if ([contentSizeCategory isEqualToString:UIContentSizeCategoryExtraExtraExtraLarge])
		p += 5;
	else if ([contentSizeCategory isEqualToString:UIContentSizeCategoryAccessibilityMedium] ||
			 [contentSizeCategory isEqualToString:UIContentSizeCategoryAccessibilityLarge] ||
			 [contentSizeCategory isEqualToString:UIContentSizeCategoryAccessibilityMedium] ||
			 [contentSizeCategory isEqualToString:UIContentSizeCategoryAccessibilityExtraLarge] ||
			 [contentSizeCategory isEqualToString:UIContentSizeCategoryAccessibilityExtraExtraLarge] ||
			 [contentSizeCategory isEqualToString:UIContentSizeCategoryAccessibilityExtraExtraExtraLarge])
		p += 5;
	return MAX(p, 11);
}

@end

File: Neocom/Neocom/NCLabel.swift
//
//  NCLabel.swift
//  Neocom
//
//  Created by Artem Shimanski on 08.12.16.
//  Copyright © 2016 Artem Shimanski. All rights reserved.
//

import UIKit

class NCLabel: UILabel {
	var pointSize: CGFloat = 15
	
	override func awakeFromNib() {
		super.awakeFromNib()
		pointSize = font.pointSize
		NotificationCenter.default.addObserver(self, selector: #selector(didChangeContentSizeCategory(_:)), name: NSNotification.Name.UIContentSizeCategoryDidChange, object: nil)
	}
	
	deinit {
		NotificationCenter.default.removeObserver(self, name: NSNotification.Name.UIContentSizeCategoryDidChange, object: nil)
	}
	
	func fontSize(contentSizeCategory: UIContentSizeCategory) -> CGFloat {
		var p = pointSize
		switch contentSizeCategory {
		case UIContentSizeCategory.extraSmall:
			p -= 2
		case UIContentSizeCategory.small:
			p -= 1
		case UIContentSizeCategory.large:
			p += 1
		case UIContentSizeCategory.extraLarge:
			p += 3
		case UIContentSizeCategory.extraExtraLarge:
			p += 5
		case UIContentSizeCategory.extraExtraExtraLarge:
			p += 5
		case UIContentSizeCategory.accessibilityMedium,
		     UIContentSizeCategory.accessibilityLarge,
			UIContentSizeCategory.accessibilityExtraLarge,
			UIContentSizeCategory.accessibilityExtraExtraLarge,
			UIContentSizeCategory.accessibilityExtraExtraExtraLarge:
			p += 5
		default:
			break
		}
		
		return max(p, 11)
	}
	
	func didChangeContentSizeCategory(_ note: NSNotification) {
		font = font.withSize(fontSize(contentSizeCategory: UIApplication.shared.preferredContentSizeCategory))
		invalidateIntrinsicContentSize()
	}
}
